RTClib.cpp

```diff
@@ -746,6 +746,168 @@ TimeSpan::TimeSpan(int16_t days, int8_t hours, int8_t minutes, int8_t seconds)
 /**************************************************************************/
 TimeSpan::TimeSpan(const TimeSpan &copy) : _seconds(copy._seconds) {}
 
+/*!
+    @brief  Create a new TimeSpan from an iso8601 formatted period string
+
+    If the string is entirely malformed (doesn't begin with P or -) this will
+    construct a TimeSpan of 0 seconds. If parsing fails before the end of the
+    string, this constuctor will interpret its argument as the longest
+    well-formed prefix string.
+
+    For example, the string P5M10~ will parse the same as P5M
+
+    @param iso8601 the formatted string
+*/
+TimeSpan::TimeSpan(char *iso8601) : _seconds(0) {
+  int32_t sign = 1;
+  char *cursor = iso8601;
+  if (*cursor == '-') {
+    sign = -1;
+    cursor++;
+  }
+  if (*cursor == 'P') {
+    cursor++;
+  } else {
+    return;
+  }
+
+  char *rest;
+  long num = strtol(cursor, &rest, 10);
+  cursor = rest;
+  if (*cursor == 'D') {
+    _seconds += sign * SECONDS_PER_DAY * num;
```

Comment:
This can overflow.

Reply:
Yeah, it can. What should I do about it? Detect overflow, return 0 on failure, and add this failure mode to the docs?

Reply:
I'm not sure. It seems to me that the overflow cannot be detected without significant extra complexity. :-(

Reply:
How would you feel about using [`__builtin_smull_overflow`](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html). It is supported in both gcc and clang, but it's non-standard.

Reply:
`__builtin_smull_overflow()` looks nice, but I do not know whether it is supported by all the environments where RTClib can potentially be used.

Maybe a simple fix would be to build the absolute value of the duration as a `uint32_t`. That won't prevent the overflow, but at least it would avoid undefined behavior. Then, at the end of the constructor,

```c++
_seconds = sign * abs_seconds;
```

is guaranteed to wrap correctly to the correct signed value, as long as the result is in the range of an `int32_t`.

Reply:
> all the environments where RTClib can potentially be used

Is RTClib used outside the Arduino ecosystem? And if so, can it be used with any compiler other than Clang or GCC? (clang [includes these intrinsics ](https://clang.llvm.org/docs/LanguageExtensions.html#multiprecision-arithmetic-builtins) too)